---
title: "Analysis of BCR"
author: "Dianyu Chen"
date: "2021/4/7"
output: 
  html_document:
      toc: true
      theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
  
In this vignette, we show how to combine BCR information with scRNA-seq data to find difference between high- and low-affinity cells. To keep the vignette concise, we will use pre-built BCR annotation file and Seurat object with full meta information.  
**Note:** detailed codes to parse results of CellRanger VDJ and IgBlast to get the information about mutation and affinity are in another script: `Igblast_parsing.Rmd`. If you are interested in how we perform the analysis, you can check that file.  
  
```{r, message=FALSE, warning=FALSE}
library(Seurat)
library(tidyverse)
library(RColorBrewer)
```

## V gene usage
Based on the parsing result of IgBlast, we can plot V gene usage of NP-specific B cells. In consistent with our expectation, V gene segment `IGHV1-72*01 (VH186.2)` is the most abundant one.
```{r}
# BCR analysis
bcr_batch1 <- read.delim("../data/gcb_np_batch1_igblast_result.tsv",
                         stringsAsFactors = FALSE)
bcr_clean_batch1 <- bcr_batch1[bcr_batch1$Vgene_match==1, ]
bcr_clean_batch1$barcode <- paste0("gcb_np_batch1_", 
                                   str_split(bcr_clean_batch1$barcode, "_", simplify = TRUE)[,1])

bcr_batch2 <- read.table("../data/gcb_np_batch2_igblast_result.tsv",
                         sep = "\t",
                         header = TRUE,
                         stringsAsFactors = FALSE)
bcr_clean_batch2 <- bcr_batch2[bcr_batch2$Vgene_match==1, ]
bcr_clean_batch2$barcode <- paste0("gcb_np_batch2_", 
                                   str_split(bcr_clean_batch2$barcode, "_", simplify = TRUE)[,1])
bcr_clean_all <- rbind(bcr_clean_batch1, bcr_clean_batch2)

## Donut plot of IGHV usage
bcr_clean_all %>%
    group_by(v_gene) %>%
    summarise(count = n()) %>%
    mutate(fraction = count/sum(count),
           ymax = cumsum(fraction),
           ymin = c(0, head(ymax, n=-1))) %>%
    ggplot(aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=v_gene)) +
    geom_rect(color="black") +
    coord_polar(theta="y") + 
    xlim(c(2, 4)) +
    scale_fill_manual(values = colorRampPalette(brewer.pal(12, "Paired"))(61)) +
    theme_void()
```

## Mutation landscape
To show mutation landscape of IGHV1-72*01 across V gene, we need to use raw IgBlast result directly. 
```{r, warning=FALSE, message=FALSE}
## parsing igblast_vh186.2 result
igblast_batch1 <- read.delim("../data/gcb_np_batch1.tsv", stringsAsFactors = FALSE)
igblast_batch1$sequence_id <- paste0("gcb_np_batch1_", igblast_batch1$sequence_id)
igblast_batch2 <- read.delim("../data/gcb_np_batch2.tsv", stringsAsFactors = FALSE)
igblast_batch2$sequence_id <- paste0("gcb_np_batch2_", igblast_batch2$sequence_id)
igblast <- bind_rows(igblast_batch1, igblast_batch2) 

## vgene mutation position offset
igblast_vh186.2 <- igblast[igblast$v_call=="IGHV1-72*01", ]
vgene_mutation_offset <- (igblast_vh186.2$v_germline_start + 1) %/% 3

## compare each amino acid in sequence
vgene_aa_match <- sapply(1:nrow(igblast_vh186.2), function(x) {
    apply(do.call(rbind, strsplit(c(igblast_vh186.2[x,]$v_sequence_alignment_aa, igblast_vh186.2[x,]$v_germline_alignment_aa), "")), 
          2, 
          function(i){
              i[1] == i[2]
          }
    )
}
)
vgene_mutation_pos <- lapply(vgene_aa_match, function(x) which(!x))
## add offset for position
vgene_mutation_pos <- lapply(1:nrow(igblast_vh186.2), function(i) vgene_mutation_pos[[i]] + vgene_mutation_offset[i])

mutation_pos <- data.frame(pos=unlist(vgene_mutation_pos))
             
## Is there any more elegant way to add labels?
ggplot(mutation_pos, aes(pos)) +
    geom_bar(aes(y = (..count..)/length(vgene_mutation_pos)), fill = "#3C548899") + 
    scale_x_continuous(limits = c(0,99),
                       breaks = seq(0,100,1),
                       labels = c(0, rep("",9), 10, rep("",9), 20, rep("",9), 30,
                                  rep("",9), 40, rep("",9), 50, rep("",9), 60,
                                  rep("",9), 70, rep("",9), 80, rep("",9), 
                                  90, rep("",9), 100)) +
    theme(panel.background = element_rect(fill = "white"),
          axis.line = element_line(colour = "black"),
          axis.ticks.x = element_line(size = 1), 
          axis.ticks.length=unit(.25, "cm")) +
    coord_cartesian(expand = FALSE) +
    ylab("Amino acid substitution")
```

We can also plot distribution of amino acid at position 33 of IGHV1-72*01. Except the germline one `W`, the most abundant mutation is `L`, in line with our expectations.
```{r, warning=FALSE, message=FALSE}
as.data.frame(table(str_split(igblast_vh186.2$v_sequence_alignment_aa, "", simplify = T)[, 33])) %>%
    arrange(Freq) %>% 
    mutate(Var1 = factor(Var1, levels = Var1)) %>% 
    mutate(fraction = Freq/sum(Freq),
           ymax = cumsum(fraction),
           ymin = c(0, head(ymax, n=-1))) %>%
    ggplot(aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=Var1)) +
    geom_rect(color = "black") +
    coord_polar(theta="y") + 
    xlim(c(2, 4)) +
    theme_void() +
    scale_fill_manual(values = rev(brewer.pal(12, "Paired")[1:11]))
```
  
## Comparison between high- and low-affinity group
* Topic
```{r}
## e.g. Topic 1
gcb_np_integrated <- readRDS("../data/gcb_np_integrated_20200908.RDS")
gcb_np_integrated@meta.data %>%
    filter(affinity %in% c("High_affinity", "Low_affinity")) %>%
    mutate(affinity=factor(affinity, levels = c("Low_affinity", "High_affinity"))) %>% 
    ggplot(aes(affinity, Topic_1, color = affinity)) +
    geom_boxplot(size=1) +
    scale_color_manual(values = c("#3C5488FF", "#E64B35FF")) +
    theme(panel.background = element_rect(fill = "white"),
          axis.line = element_line(colour = "black"),
          axis.ticks.x = element_line(size = 1))

t.test(Topic_1~affinity, 
       gcb_np_integrated@meta.data %>% filter(affinity %in% c("High_affinity", "Low_affinity")))
```

* Signature
```{r}
## e.g. OXPHOS
gcb_np_integrated@meta.data %>%
    filter(affinity %in% c("High_affinity", "Low_affinity")) %>%
    mutate(affinity=factor(affinity, levels = c("Low_affinity", "High_affinity"))) %>% 
    ggplot(aes(affinity, oxp_score, color = affinity)) +
    geom_boxplot(size=1) +
    scale_color_manual(values = c("#3C5488FF", "#E64B35FF")) +
    theme(panel.background = element_rect(fill = "white"),
          axis.line = element_line(colour = "black"),
          axis.ticks.x = element_line(size = 1))

t.test(oxp_score~affinity,
       gcb_np_integrated@meta.data %>% filter(affinity %in% c("High_affinity", "Low_affinity")))
```

## Session info
```{r}
sessionInfo()
```
